aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -48,13 +48,87 @@ namespace detail {
     }
   };
 
+  struct TensorOptionsAccumulator : at::IterArgs<TensorOptionsAccumulator> {
+    TensorOptions options;
+    void operator()(c10::optional<ScalarType> dtype) {
+      if (dtype.has_value()) {
+        options = options.dtype(*dtype);
+      }
+      else {
+        // Following the TensorOptions logic where in case of undefined
+        // dtype, default dtype is returned.
+        // Issue https://github.com/pytorch/pytorch/issues/30405
+        options = options.dtype(at::get_default_dtype());
+      }
+    }
+    void operator()(c10::optional<Device> device) {
+      if (device.has_value()) {
+        options = options.device(*device);
```

Comment:
I'm pretty sure the setters all take `optional`, so you can just bang `device` in directly here without doing a test here...